lib/sycamore/sycamore/writers/weaviate_writer.py

```diff
@@ -0,0 +1,120 @@
+import logging
+from typing import Any, Iterable, Optional
+
+from ray.data import Dataset, Datasink
+from ray.data._internal.delegating_block_builder import DelegatingBlockBuilder
+from ray.data._internal.execution.interfaces import TaskContext
+from ray.data.block import Block, BlockAccessor
+from sycamore.data.document import Document
+from sycamore.plan_nodes import Node, Write
+from weaviate import WeaviateClient
+from weaviate.collections.classes.data import DataReference
+
+
+class WeaviateWriter(Write):
+    def __init__(
+        self,
+        plan: Node,
+        collection_name: str,
+        client_params: dict,
+        collection_config: Optional[dict],
+        **ray_remote_args,
+    ):
+        super().__init__(plan, **ray_remote_args)
+        self.collection_name = collection_name
+        self.client_params = client_params
+        self.collection_config = collection_config
+
+    def execute(self) -> Dataset:
+        dataset = self.child().execute()
+        datasink = WeaviateDatasink(self.collection_name, self.client_params, self.collection_config)
+        dataset.write_datasink(datasink, ray_remote_args=self.resource_args)
+        return dataset
+
+
+class WeaviateDatasink(Datasink):
+    def __init__(self, collection_name: str, client_params: dict, collection_config: Optional[dict]):
+        self._collection_name = collection_name
+        self._client_params = client_params
+        self._collection_config = collection_config
+
+    def on_write_start(self):
+        client = WeaviateClient(**self._client_params)
+        with client:
+            if self._collection_config is not None:
+                if client.collections.exists(self._collection_name):
+                    logging.warning(
+                        f"Collection config was provided, but collection {self._collection_name} "
+                        "already exists, so ignoring provided config."
+                    )
+                else:
+                    client.collections.create(**self._collection_config)
+
+    def write(self, blocks: Iterable[Block], ctx: TaskContext) -> Any:
+        builder = DelegatingBlockBuilder()
+        for block in blocks:
+            builder.add_block(block)
+        block = builder.build()
+        client = WeaviateClient(**self._client_params)
+        with client:
+            collection = client.collections.get(self._collection_name)
+            with collection.batch.dynamic() as batch:
+                objects = self._extract_weaviate_objects(block)
+                refs = []
+                for obj in objects:
+                    if "references" in obj:
+                        obj_refs = obj.pop("references")
```

Comment:
This does two hash lookups in quick succession.  Better would be like:
```
obj_refs = obj.pop("references", None)
if obj_refs is not None:

Reply:
ack